src/Grpc.StatusProto/README.md

```diff
@@ -0,0 +1,256 @@
+
+* The server side uses C#'s Object and Collection initializer syntax. The avoids the needs to a *builder* API to be developed.
+
+## Introduction to the richer error model
+
+Google APIs define an error model that allows a server to return an error status that is more than
+just a status code or short message.  The status returned can contain a list of various details giving
+more information about the error.
+
+This error model is define by the protocol buffers files [status.proto](https://github.com/googleapis/googleapis/blob/master/google/rpc/status.proto)
+and [error_details.proto](https://github.com/googleapis/googleapis/blob/master/google/rpc/error_details.proto)
+and is implemented in classes provided in the `Google.Api.CommonProtos` NuGet package.
+
+You can use this same error model in your gRPC services. This package provides classes to support you in
+doing this. In this error model the error is encapsulated by an instance of `Google.Rpc.Status` and
+returned in the trailing response metadata.  Setting and reading this metadata is handled for you. Details below.
+
+## Server Side
+
+The server side uses C#'s Object and Collection initializer syntax. The avoids the needs to a *builder* API to be developed.
+
+There are two ways that the server can return an error status:
+- by throwing an `RpcException` that contains the details of the status.
+- by explicitly set the `Status` and `ResponseTrailers` in the `ServerCallContext`
+before returning.
+
+There are examples of both methods below.  In a .NET client the result will always be
+an exception received by the client no matter which of the above methods the server
+implements.
+
+The `Google.Rpc.Status` can be created and initialized using C#'s Object and Collection initializer syntax. To add messages to the `Details` repeated field, wrap each one in `Any.Pack()` - see example below.
+
+The `Google.Rpc.Status` extension method `ToRpcException` creates the appropriate `RpcException` from the status.
+
+The `Grpc.Core.Metadata` extension method `SetRpcStatus` adds a binary representation of the status to the metadata.
+
+
+Example - throwing a RcpException:
+```C#
+throw new Google.Rpc.Status
+{
+    Code = (int)StatusCode.NotFound,
+    Message = "Simple error message",
+    Details =
+    {
+        Any.Pack(new ErrorInfo
+        {
+            Domain = "error example",
+            Reason = "some reason"
+        }),
+        Any.Pack(new RequestInfo
+        {
+            RequestId = "EchoRequest",
+            ServingData = "Param: " + request.Action.ToString()
+        }),
+        }
+}.ToRpcException();
+```
+Example - setting the status and response trailers instead of throwing an exception:
+```C#
+context.Status = new Grpc.Core.Status(StatusCode.Internal, "Some detail");
+context.ResponseTrailers.SetRpcStatus(new Google.Rpc.Status
+{
+    Code = (int)StatusCode.NotFound,
+    Message = "Simple error message",
+    Details =
+    {
+        Any.Pack(new ErrorInfo
```

Comment:
FYI: There are Google-internal proposals around the use of ErrorInfo; I'm checking on the status of those, but they could affect the preferred layout for the richer error model.